# First Issue
while doing layered architecture and seperating test scripts to test folder,
I got to import error from parent directory.
To solve this error, I searched and knew the concept of module, package and so on.
Here is the reference doc I found: https://dojang.io/mod/page/view.php?id=2448

1) __name__ = '__main__' 의 의미
    - python 에서 첫 파일의 __name__ 은 '__main__' 이 된다
    - import aaa.py 하면 aaa.py 가 실행되는데, 여기서 __name__ 은 'aaa'가 된다
    - 즉, if __name__='__main__' 구문은 import 가 아니라 직접 실행했을 때에만 실행되는 구문을 지정하는 것이다

2) __init__.py 파일의 의미
    - 해당 디렉토리가 패키지라는 것을 명시 + 초기화 함수
    - 패키지 안의 모듈 (example.py) 을 __init__.py 에서 import 하여 외부에서 세부 모듈을 지정하지 않아도 편리하게 사용할 수도 있고,
    - "__all__" 을 이용하여 보안을 위해 공개하고자 하는 모듈을 지정할 수도 있음

3) __doc__
    - 모듈의 docstring 이다. 모듈의 첫 줄에 '''this is a doc string''' 이렇게 하면 aaa.__doc__ 하면 나온다.

4) python prompt >>> import sys, sys.path
    - 파이썬에서 현재 경로에 모듈, 패키지가 없으면 sys.path 에 있는 경로에서 모듈, 패키지를 찾는다
    - python aaa.py 로 실행하면 sys.path 에 현재 디렉토리가 포함되어 있다 
        - 즉, 현재 디렉토리의 bbb.py 는 import bbb 해서 들고 올 수 있다
        - 허나 상위 디렉토리나 하위 디렉토리의 경우에는 어려움이 있다
    - 패키지(모듈) 로 파이썬을 실행하고 상대 경로로 import 하면 해결할 수 있다
        - 만약 root/aaa.py 의 구조라면,
        - python -m root.aaa 로 모듈 옵션으로 파이썬을 실행하면,
        - import .bbb, import ..ccc 등으로 자유롭게 상대경로로 import 할 수 있다.
        - 허나 이를 사용하려면, __init__.py 를 폴더별로 추가하여 패키지라는 것을 명시해야 한다

5) python prompt >>> dir()
    - 파이썬에서 네임스페이스를 확인하는 것이다
    - import sys 하면 네임스페이스에 sys 가 추가되어 sys 안의 함수를 호출할 수 있다

문제 해결
- https://myjorney.tistory.com/52
- https://velog.io/@hamsterhamin/Python-ImportError-attempted-relative-import-with-no-known-parent-package
- pytest test_view.py 할 때에 해당 스크립트를 바로 실행하는데,
    - python interpreter 가 상대경로의 기준점을 찾지 못해 발생 (첫 실행시 __main__ 인데 이게 어디인지 모름)
    - 절대 경로로 바꾸자. --> x
- test 폴더와 루트 폴더에 __init__.py 을 추가하여 패키지임을 명시하자
    - 그리고 from 구문에 . 을 추가하여 상대경로임을 명시하자
- nohup python 할 때에 setup.py 자체를 실행하면 에러가 뜨는데,
    - 상위 폴더로 가서 'nohup python -m api_project.setup ~~ ' 이렇게
    - 모듈 옵션을 줘서 실행하자!

# Second Issue
test each layer functions (model, service, view), it occured erros whlie importing config
In config.py, test_db and test_jwt_secret_key is stored in different format.

문제 해결
- test 폴더 안의 test_service.py, test_view.py 에서 config 를 가져오는 방식을 디버깅했다.
- 특별한 이슈가 있다기보다는 복잡해서 헷갈린 것이 크다..